apps/application-system/api/src/app/modules/application/tools/templateApiActionRunner.service.ts

```diff
@@ -0,0 +1,164 @@
+        )
+        return ps.then(() => {
+          return
+        })
+      })
+    }, Promise.resolve())
+
+    await result
+  }
+
+  groupByOrder(
+    actions: ApplicationTemplateAPIAction[],
+  ): ApplicationTemplateAPIAction[][] {
+    const groupedActions = actions.reduce((acc, action) => {
+      const order = action.order ?? -1
+      if (!acc[order]) {
+        acc[order] = []
+      }
+      acc[order].push(action)
+      return acc
+    }, {} as { [key: number]: ApplicationTemplateAPIAction[] })
+
+    return Object.keys(groupedActions).map(
+      (value: string): ApplicationTemplateAPIAction[] => {
+        return groupedActions[(value as unknown) as number]
+      },
+    )
+  }
+
+  async callProvider(action: ApplicationTemplateAPIAction) {
+    const { apiModuleAction, externalDataId, mockData, namespace } = action
+    console.log(action)
+    let actionResult: PerformActionResult | undefined
+
+    const useMocks =
+      typeof action.useMockData === 'function'
+        ? action.useMockData(this.application)
+        : action.useMockData
+
+    if (useMocks) {
+      console.log('Using mock data for action', apiModuleAction)
+      actionResult =
+        typeof mockData === 'function' ? mockData(this.application) : mockData
+    } else {
+      console.log('Calling provider for action', apiModuleAction)
+      actionResult = await this.templateAPIService.performAction({
+        templateId: this.application.typeId,
+        type: apiModuleAction,
+        namespace,
+        props: {
+          application: this.application,
+          auth: this.auth,
+        },
+      })
+    }
+
+    if (!actionResult)
+      throw new Error(`No Action or mock is defined for ${apiModuleAction}`)
+
+    await this.updateExternalData(actionResult, apiModuleAction, externalDataId)
+  }
+
+  async persistExternalData(
+    actions: ApplicationTemplateAPIAction[],
+  ): Promise<void> {
+    actions.map((action) => {
+      if (!action.shouldPersistToExternalData) {
+        delete this.newExternalData[
+          action.externalDataId || action.apiModuleAction
+        ]
+      }
+    })
+
+    await this.applicationService.updateExternalData(
+      this.application.id,
+      this.oldExternalData,
+      this.newExternalData,
+    )
+  }
+
+  async updateExternalData(
+    actionResult: PerformActionResult,
+    apiModuleAction: string,
+    externalDataId?: string,
+  ): Promise<void> {
+    const newExternalDataEntry: ExternalData = {
+      [externalDataId || apiModuleAction]: {
+        status: actionResult.success ? 'success' : 'failure',
+        date: new Date(),
+        data: actionResult.success
+          ? (actionResult.response as ExternalData['data'])
+          : actionResult.error,
+      },
+    }
+
+    this.newExternalData = { ...this.newExternalData, ...newExternalDataEntry }
+
+    this.application.externalData = {
+      ...this.application.externalData,
+      ...this.newExternalData,
```

Comment:
My thought was not to override if the application has on some previous step updated the external data.

Reply:
Sure, I guess this is just an optimization comment:

```
{
  ...this.application.externalData,
  ...this.newExternalData,
}
// is basically the same but slower than:
{
  ...this.application.externalData,
  ...this.newExternalDataEntry,
}
// because you've already assigned previous entries to externalData, and you are still cloning all those old changes.
```